File: problems/1239_maximum_length_of_a_concatenated_string_with_unique_characters/Solution.java
import java.util.*;

class Solution {
    public int maxLength(List<String> arr) {
        List<Integer> dp = new ArrayList<>();
        dp.add(0);
        int res = 0;
        for (String s : arr) {
            int mask = 0; boolean dup = false;
            for (char c : s.toCharArray()) {
                int bit = 1 << (c - 'a');
                if ((mask & bit) != 0) { dup = true; break; }
                mask |= bit;
            }
            if (dup) continue;
            List<Integer> newDp = new ArrayList<>();
            for (int existing : dp) {
                if ((existing & mask) != 0) continue;
                int combined = existing | mask;
                newDp.add(combined);
                res = Math.max(res, Integer.bitCount(combined));
            }
            dp.addAll(newDp);
        }
        return res;
    }
}

File: problems/1239_maximum_length_of_a_concatenated_string_with_unique_characters/solution.py
class Solution:
    def maxLength(self, arr: list[str]) -> int:
        dp = [set()]
        for s in arr:
            if len(set(s)) != len(s):
                continue
            cs = set(s)
            new_dp = []
            for existing in dp:
                if cs & existing:
                    continue
                new_dp.append(cs | existing)
            dp.extend(new_dp)
        return max(len(x) for x in dp)
